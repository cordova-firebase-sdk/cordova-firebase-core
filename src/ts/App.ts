import { Promise } from "es6-promise";
import { execCmd } from "./CommandQueue";
import { isInitialized } from "./common";
import { PluginBase } from "./PluginBase";

export interface IAppInitializeOptions {

  /**
   * Auth / General Use
   */
  apiKey?: string;

  /**
   * Auth with popup/redirect
   */
  authDomain?: string;

  /**
   * Realtime Database
   */
  databaseURL?: string;

  /**
   * Storage
   */
  storageBucket?: string;

  /**
   * Cloud Messaging
   */
  messagingSenderId?: string;
}

export class App extends PluginBase {

  public name: string = "[DEFAULT]";

  private _options: IAppInitializeOptions = {
    apiKey: null,
    authDomain: null,
    databaseURL: null,
    messagingSenderId: null,
    storageBucket: null,
  };

  private _isInitialized: boolean;

  /**
   * @constructor
   * @param [name] - Application name. Default value is "[DEFAULT]"
   * @param [initOptions] - Application initialize option.
   *  This parameter is for {@link: plugin.firebase.initializeApp()} method.
   */
  constructor(name?: string, initOptions?: IAppInitializeOptions) {
    super("fireapp");

    this._isReady = true;
    this._options = {
      apiKey: initOptions.apiKey || null,
      authDomain: initOptions.authDomain || null,
      databaseURL: initOptions.databaseURL || null,
      messagingSenderId: initOptions.messagingSenderId || null,
      storageBucket: initOptions.storageBucket || null,
    };

    execCmd({
      args: [{
        id: this.id,
        name: this.name,
        options: this._options,
      }],
      context: this,
      methodName: "newInstance",
      pluginName: "CordovaFirebaseCore",
    }).then(() => {
      this._isInitialized = true;
      this._trigger("ready");
    });
  }

  /**
   * Generate FirebaseDatabse instance
   * @param [url] - Realtime database url
   * @returns database instance
   */
  public database(url?: string): any {
    if (isInitialized("plugin.firebase.database._DBs")) {
      const CordovaFirebaseDatabase: any = require("cordvoa-firebase-datbase.cordova-firebase-database");
      const options: any = Object.create(this._options);
      if (url) {
        options.databaseURL = url;
      }
      const db = new CordovaFirebaseDatabase(this.id, options);
      if (this._isInitialized) {
        db._trigger("fireAppReady");
      } else {
        this._on("ready", () => {
          db._trigger("fireAppReady");
        });
      }
    } else {
      throw new Error("cordova-firebase-databse plugin is required");
    }
  }
}
