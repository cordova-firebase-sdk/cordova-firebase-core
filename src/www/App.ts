// import { Promise } from "es6-promise";
import { exec } from "cordova";
import { isInitialized } from "./common";
import { IAppInitializeOptions } from "./IAppInitializeOptions";
import { PluginBase } from "./PluginBase";

declare let Promise: any;
declare let window: any;

export class App extends PluginBase {

  public name: string = "[DEFAULT]";

  private _options: IAppInitializeOptions = {
    apiKey: null,
    authDomain: null,
    databaseURL: null,
    messagingSenderId: null,
    storageBucket: null,
  };

  /**
   * @constructor
   * @param [name] - Application name. Default value is "[DEFAULT]"
   * @param [initOptions] - Application initialize option.
   *  This parameter is for {@link: plugin.firebase.initializeApp()} method.
   */
  constructor(name?: string, initOptions?: IAppInitializeOptions) {
    super("fireapp");

    if (typeof name === "string") {
      this.name = name;
    }

    if (initOptions) {
      this._options = {
        apiKey: initOptions.apiKey || null,
        authDomain: initOptions.authDomain || null,
        databaseURL: initOptions.databaseURL || null,
        messagingSenderId: initOptions.messagingSenderId || null,
        storageBucket: initOptions.storageBucket || null,
      };
    }

    // Create one new instance in native side.
    exec(() => {
      this._isReady = true;
      this._trigger("ready");
    },
    (error: any) => {
      throw new Error(error);
    },
    "CordovaFirebaseCore",
    "newInstance",
    [{
      id: this.id,
      name: this.name,
      options: this._options,
    }]);
  }

  /**
   * Generate FirebaseDatabse instance
   * @param [url] - Realtime database url
   * @returns database instance
   */
  public database(url?: string): any {
    const options: any = this.options;
    if (url) {
      options.databaseURL = url;
    }

    // Load `cordova-firebase-database.Database` module if not yet.
    let database: any;
    if (isInitialized("plugin.firebase.database") &&
      typeof window.plugin.firebase.database === "function") {
      database = window.plugin.firebase.database(this, options);
    } else {
      const moduleName: string = "cordova-firebase-database";

      if (!window.cordova.define.moduleMap[moduleName + ".Database"]) {
        // cordova-firebase-database is not installed.
        throw new Error(moduleName + " plugin is required");
      }
      cordova.require(moduleName + ".Database");
      database = window.plugin.firebase.database(this, options);
    }

    // Waits if native side of FirebaseAppPlugin is not ready yet.
    if (this._isReady) {
      database._trigger("fireAppReady");
    } else {
      this._one("ready", () => {
        database._trigger("fireAppReady");
      });
    }
    return database;
  }

  public get options(): IAppInitializeOptions {
    return JSON.parse(JSON.stringify(this._options));
  }
}


class SecretAppManager {

  public get apps(): Array<App> {
    const keys: Array<string> = Object.keys(this._apps);
    const apps: Array<App> = [];
    keys.forEach((key: string) => {
      apps.push(this._apps[key]);
    });
    return apps;
  }

  public _apps: any = {};

  /**
   * @param initOptions - Application initialize options
   * @param [name] - Application name.
   * @returns Application instance
   */
  public initializeApp(initOptions: IAppInitializeOptions, name?: string): App {
    name = name || "[DEFAULT]";
    if (name in this._apps) {
      throw new Error("Name '" + name + "' application has been already existed.");
    }
    const app: App = new App(name, initOptions);
    this._apps[name] = app;
    return app;
  }
}
if (window.cordova && window.cordova.version) {
  const manager: SecretAppManager = new SecretAppManager();

  const firebaseNS: any = {
    apps: manager.apps,

    ANDROID_SDK_VERSION: "5.5.0",

    database: undefined,

    initializeApp: manager.initializeApp.bind(manager),

    IOS_SDK_VERSION: "5.5.0",

    Promise: Promise.class,

    WEBJS_SDK_VERSION: "5.5.0",
  };

  window.cordova.addConstructor(() => {
    window.plugin = window.plugin || {};
    // (window as any).plugin.firebase = (window as any).plugin.firebase || {};
    if (!window.plugin.firebase) {
      Object.defineProperty(window.plugin, "firebase", {
        value: firebaseNS,
      });
      Object.defineProperty(window.plugin.firebase, "app", {
        value: (name?: string): App => {
          name = name || "[DEFAULT]";
          const app: App = manager._apps[name];
          if (app) {
            return app;
          } else {
            throw new Error("Default app has been not initialized.");
          }
        },
      });
    }
  });
}
