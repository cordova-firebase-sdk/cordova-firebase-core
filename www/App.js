







var utils = require('cordova/utils'),
  common = require('./Common'),
  BaseClass = require('./BaseClass'),
  execCmd = require('./FirebaseCoreCommandQueue');

/*******************************************************************************
 * @name App
 ******************************************************************************/
function App(options, name) {
  BaseClass.call(this);

  var self = this;

  Object.defineProperty(self, 'options', {
    value: options
  });
  Object.defineProperty(self, 'name', {
    value: name
  });

  Object.defineProperty(self, 'id', {
    value: self.hashCode + '_fireapp',
    writable: true,
    enumerable: false
  });
  Object.defineProperty(self, '_isReady', {
    value: false,
    writable: true,
    enumerable: false
  });

  Object.defineProperty(self, '_isRemoved', {
    value: false,
    enumerable: false
  });

  execCmd.call({
    '_isReady': true
  }, function() {
    self._isReady = true;
    self._trigger('ready');
  }, function(error) {
    if (error instanceof Error) {
      throw error;
    } else {
      throw new Error(error);
    }
  }, 'CordovaFirebaseCore', 'newInstance', [{
    'id': self.id,
    'options': options,
    'name': name
  }], {
    'sync': true
  });
}

utils.extend(App, BaseClass);


//---------------------------------------------------------------------------------
// App.database
// https://firebase.google.com/docs/reference/js/firebase.app.App#database
//---------------------------------------------------------------------------------
App.prototype.database = function(url) {
  var self = this;
  if (common.isInitialized('plugin.firebase.database._DBs')) {
    var CordovaFirebaseDatabase = require('cordova-firebase-database.cordova-firebase-database');
    var options = utils.clone(this.options);
    if (url) {
      options.databaseURL = url;
    }
    var db = new CordovaFirebaseDatabase(self.id, options);
    if (self._isReady) {
      db._trigger('fireAppReady');
    } else {
      self._on('ready', function() {
        db._trigger.call(db, 'fireAppReady');
      });
    }
    window.plugin.firebase.database._DBs[db.id] = db;
    return db;
  } else {
    throw new Error('cordova-firebase-database plugin is required to use this feature.');
  }
};







module.exports = App;
