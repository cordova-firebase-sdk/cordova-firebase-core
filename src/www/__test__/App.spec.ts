import { App } from "../App";

describe("[App]", () => {

  it ("name should be [DEFAULT]", () => {
    const app: App = new App();
    expect(app.name).toEqual("[DEFAULT]");
  });
  it ("name should be 'hello'", () => {
    const app: App = new App("hello");
    expect(app.name).toEqual("hello");
  });
});
