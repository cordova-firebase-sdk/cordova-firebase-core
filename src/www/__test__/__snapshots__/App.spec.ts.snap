// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[App] [App.databaseURL option] should throw error if does not point the root reference 1`] = `"Database URL must point to the root of a Firebase Database (not including a child path). "`;

exports[`[App] [App.databaseURL option] should throw error if does not start with https://(projectId).firebaseio.com 1`] = `"Cannot parse Firebase url. Please use https://<YOUR FIREBASE>.firebaseio.com"`;
